File: project/src/main/java/com/expedia/bookings/location/CurrentLocationSuggestionProvider.java
package com.expedia.bookings.location;

import java.util.List;

import android.location.Location;

import com.expedia.bookings.data.cars.ApiError;
import com.expedia.bookings.data.cars.Suggestion;
import com.expedia.bookings.data.pos.PointOfSale;
import com.expedia.bookings.services.SuggestionServices;

import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func1;

public class CurrentLocationSuggestionProvider {

	SuggestionServices suggestionServices;
	Observable<Location> locationObservable;

	public CurrentLocationSuggestionProvider(SuggestionServices suggestionServices, Observable<Location> locationObservable) {
		this.suggestionServices = suggestionServices;
		this.locationObservable = locationObservable;
	}

	public Observable<Suggestion> currentLocationSuggestion() {
		return locationObservable.flatMap(new LocationToSuggestions(ENSURE_ATLEAST_ONE_SUGGESTION));
	}

	protected class LocationToSuggestions implements Func1<Location, Observable<Suggestion>> {

		Action1<List<Suggestion>> validateSuggestionList;
		LocationToSuggestions(Action1<List<Suggestion>> validateSuggestionList) {
			this.validateSuggestionList = validateSuggestionList;
		}

		@Override
		public Observable<Suggestion> call(Location location) {
			String latlong = "" + location.getLatitude() + "|" + location.getLongitude();

			return CurrentLocationSuggestionProvider.this.suggestionServices
				.getNearbyLxSuggestions(PointOfSale.getSuggestLocaleIdentifier(), latlong,
					PointOfSale.getPointOfSale().getSiteId())
				.doOnNext(validateSuggestionList)
				.map(TAKE_FIRST_SUGGESTION);
		}
	};

	protected static final Action1<List<Suggestion>> ENSURE_ATLEAST_ONE_SUGGESTION = new Action1<List<Suggestion>>() {
		@Override
		public void call(List<Suggestion> suggestions) {
			if (suggestions == null || suggestions.size() < 1) {
				throw new ApiError(ApiError.Code.SUGGESTIONS_NO_RESULTS);
			}
		}
	};

	private static final Func1<List<Suggestion>, Suggestion> TAKE_FIRST_SUGGESTION = new Func1<List<Suggestion>, Suggestion>() {
		@Override
		public Suggestion call(List<Suggestion> suggestions) {
			return suggestions.get(0);
		}
	};
}

File: project/src/main/java/com/expedia/bookings/location/CurrentLocationSuggestionProvider.kt
package com.expedia.bookings.location

import android.content.Context
import android.location.Location
import com.expedia.bookings.data.ApiError
import com.expedia.bookings.data.GaiaSuggestionRequest
import com.expedia.bookings.data.SuggestionV4
import com.expedia.bookings.data.pos.PointOfSale
import com.expedia.bookings.services.SuggestionV4Services
import com.expedia.bookings.utils.SuggestionV4Utils
import io.reactivex.Observable

class CurrentLocationSuggestionProvider(val suggestionServices: SuggestionV4Services, val locationObservable: Observable<Location>, val context: Context) {
    fun currentLocationSuggestion(): Observable<SuggestionV4> {
        return locationObservable.flatMap { location ->
            getNearBysuggestions(location)
                    .doOnNext { suggestions -> if (suggestions.size < 1) throw ApiError(ApiError.Code.SUGGESTIONS_NO_RESULTS) }
                    .map { suggestions ->
                        suggestions[0]
                    }
        }
    }

    fun getNearBysuggestions(location: Location): Observable<MutableList<SuggestionV4>> {
        val request = GaiaSuggestionRequest(location.latitude, location.longitude, "distance", "lx",
                PointOfSale.getSuggestLocaleIdentifier(), PointOfSale.getPointOfSale().siteId,
                misForRealWorldEnabled = false)
        return this@CurrentLocationSuggestionProvider.suggestionServices
                .suggestNearbyGaia(request)
                .map { it ->
                    SuggestionV4Utils.convertToSuggestionV4(it)
                }
    }
}
